### PR TITLE
Fix bug in window function count should count anything non-null

### LINKED ISSUE #12541: BUG: window function count should count anything non-null
Right now `np.Inf` and non-numeric types are not counted.
#### Code Sample, a copy-pastable example if possible

```
import pandas as pd
import numpy as np
from pandas import DataFrame

df = DataFrame({'A': pd.date_range('20130101',periods=4), 'B': [0, 1, 2, np.Inf], 'C':['a', 'b', 'c', 'd']})

roll = df.rolling(window=2)
res = roll.count()
print(res)
## partial output A, B don't error
     A    B
0  1.0  1.0
1  2.0  2.0
2  2.0  2.0
3  2.0  1.0
## column C errors
```
#### Expected Output

```
   A  B  C
0  1  1  1
1  2  2  2
2  2  2  2
3  2  2  2
```
#### output of `pd.show_versions()`

```
INSTALLED VERSIONS
------------------
commit: a0aaad98e8c567a3a04aef165cdb159fbd98d13f
python: 3.5.1.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.0-76-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.0rc1+77.ga0aaad9.dirty
nose: 1.3.7
pip: 8.0.1
setuptools: 19.4
Cython: 0.23.4
numpy: 1.10.2
scipy: None
statsmodels: None
xarray: None
IPython: 4.0.0
sphinx: None
patsy: None
dateutil: 2.4.2
pytz: 2015.7
blosc: None
bottleneck: 1.0.0
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.10
pymysql: None
psycopg2: None
jinja2: None
boto: None
```

I think it can be patched, together with #12538, by

```
--- a/pandas/core/window.py
+++ b/pandas/core/window.py
@@ -509,19 +509,9 @@ class _Rolling_and_Expanding(_Rolling):
         blocks, obj = self._create_blocks(how=None)
         results = []
         for b in blocks:
-
-            if com.needs_i8_conversion(b.values):
-                result = b.notnull().astype(int)
-            else:
-                try:
-                    result = np.isfinite(b).astype(float)
-                except TypeError:
-                    result = np.isfinite(b.astype(float)).astype(float)
-
-                result[pd.isnull(result)] = 0
-
-            result = self._constructor(result, window=window, min_perio
-                                       center=self.center).sum()
+            result = self._constructor(
+                b.notnull(), window=window, min_periods=0, center=self.
+            ).sum().astype(int)

```

### PR DESCRIPTION
**Pull request by mralgos**
- [x] closes #12541 
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
this looks right, pls add a note in whatsnew, bug fix section.

note that I did try to change this a while back and got some tests failures (which I think you mentioned), so these need to be investigated / fixed. I think the 'floatifying' is slightly different that checking for non-null, but need to drill down and see exactly.

File: pandas/tests/test_window.py [hunk 1]
```diff
@@ -207,6 +207,26 @@ def f():
             'A', 'ra', 'std'), ('B', 'rb', 'mean'), ('B', 'rb', 'std')])
         tm.assert_frame_equal(result, expected, check_like=True)
 
+    def test_count_nonnumeric_types(self):
+        # GH12541
+        df_inf = DataFrame({'x': [1, 2, 3], 'y': [1., 2., np.Inf]})
+        df_date = DataFrame({'x': [1, 2, 3], 
+                            'y': pd.date_range('20130101',periods=3)})
```
Comment (jreback) [inline pandas/tests/test_window.py]: can you add some NaT in here for the datetimes,  add in a timedelta example, and a float example with np.nan.

Comment (jreback) [inline pandas/tests/test_window.py]: can you do all of these with a single df (with multiple columns), and a single comparision tests (with the results)

Comment (jreback) [inline pandas/tests/test_window.py]: https://github.com/pandas-dev/pandas/pull/15054/files#diff-d1a05a7cf744e0feb0fc6cd7128903a8 is an example of how to create these various types (though need to add some nans)

Comment (mralgos) [inline pandas/tests/test_window.py]: NaN and NaT won't be counted. Is that right?

File: doc/source/whatsnew/v0.20.0.txt [hunk 2]
```diff
@@ -447,7 +447,7 @@ Bug Fixes
 
 
 - Bug in ``Series`` constructor when both ``copy=True`` and ``dtype`` arguments are provided (:issue:`15125`)
-- Bug in ``pd.read_csv()`` for the C engine where ``usecols`` were being indexed incorrectly with ``parse_dates`` (:issue:`14792`)
+- Bug in ``pd.read_csv()`` for the C engine where ``usecols`` were being indexed incorrectly with ``parse_dates`` (:issue:`14792
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.20.0.txt]: You removed something here by accident I think

File: pandas/tests/test_window.py [hunk 3]
```diff
@@ -207,6 +207,38 @@ def f():
...
+            {'int': [1, 2, 3],
+             'float': [4., 5., 6.],
+             'string': list('abc'),
+             'datetime': pd.date_range('20170101', periods=3),
+             'timedelta': pd.timedelta_range('1 s', periods=3, freq='s'),
+             'fl_inf': [1., 2., np.Inf],
+             'fl_nan': [1., 2., np.NaN],
+             'str_nan': ['aa', 'bb', np.NaN],
+             'dt_nat': [pd.Timestamp('20170101'), pd.Timestamp('20170203'),
+                        pd.Timestamp(None)]},
```
Comment (jorisvandenbossche) [inline pandas/tests/test_window.py]: Do we also need to test Periods?

**Review by jorisvandenbossche — APPROVED** [review]

File: doc/source/whatsnew/v0.20.0.txt (hunk 2 above)
Comment (mralgos) [inline doc/source/whatsnew/v0.20.0.txt]: Ops, sorry about that!

File: pandas/tests/test_window.py (hunk 3 above)
Comment (jreback) [inline pandas/tests/test_window.py]: sure could add periods in here as well.

Comment (mralgos) [inline pandas/tests/test_window.py]: Something like this?

`df = DataFrame(`
`            {...,`
`            'periods': [pd.Period('2012-01'), pd.Period('2012-02'), `
`                       pd.Period('2012-03')],`
`             ...`
`             'periods_nat': [pd.Period('2012-01'), pd.Period('2012-02'), `
`                         pd.Period(None)]},`
`            columns=cols)`

Comment (jreback) [inline pandas/tests/test_window.py]: yep

**Comment by jreback** [comment]
lgtm. ping on green (I will fix the minor typo on merge)

**Comment by jreback** [comment]
thanks @mralgos

**Pull request closed without merging**